src/test/java/io/wgj/ctci/StackTest.java

```diff
@@ -2,49 +2,121 @@
 
 import static org.junit.Assert.*;
 
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.ExpectedException;
 
-import java.util.ArrayList;
-import java.util.Collections;
 
 public class StackTest {
+    @Rule
+    public final ExpectedException thrown = ExpectedException.none();
+
     @Test
     public void testStackOneInt() {
-        Stack s = new Stack();
+        Stack<Integer> s = new Stack<>();
         s.push(1);
         int r = s.pop();
         assertEquals(1, r);
     }
 
+    @Test
+    public void testStackOneString() {
+        Stack<String> s = new Stack<>();
+        s.push("one");
+        String r = s.pop();
+        assertEquals("one", r);
+    }
+
+    @Test
+    public void testPushPopsIntermixed() {
+        Stack<Integer> s = new Stack<>();
+        s.push(0);
+        assertEquals(0, (int) s.pop());
+        s.push(1);
+        s.push(2);
+        s.push(3);
+        assertEquals(3, (int) s.pop());
+        assertEquals(2, (int) s.pop());
+        s.push(4);
+        assertEquals(4, (int) s.pop());
+        assertEquals(1, (int) s.pop());
+    }
+
     @Test
     public void testStackMultipleInts() {
-        ArrayList<Integer> expected = new ArrayList<>();
-        for (int i = 2; i < 10; i = i + 2) {
-            expected.add(i);
-        }
+        Stack<Integer> s = new Stack<>();
+        s.push(1);
+        s.push(2);
+        s.push(3);
 
-        Stack s = new Stack();
-        for (int e : expected) {
-            s.push(e);
-        }
+        assertEquals(3, (int) s.pop());
+        assertEquals(2, (int) s.pop());
+        assertEquals(1, (int) s.pop());
+    }
+
+    @Test
+    public void testGrowth() {
+        Stack<Integer> s = new Stack<>(1);
+        int startingCapacity = s.capacity();
+        s.push(1);
+        s.push(2);
+        s.push(3);
+        int newCapacity = s.capacity();
+        assertTrue(startingCapacity < newCapacity);
+
+    }
 
-        Collections.reverse(expected);
-        for (int e : expected) {
-            assertEquals(e, s.pop());
+    @Test
+    public void testGrowthWithLeftOver() {
+        int cap = 257; // 257 << 2 is 1028, and trigger a "left over".
+        Stack<Integer> s = new Stack<>(cap);
+        assertEquals(cap, s.capacity());
+        for (int i = 0; i < cap + 1; i++) {
+            s.push(0);
         }
+        // Our new capacity will use what's left over from s.grow().
+        assertEquals(s.maxSize(), s.capacity());
     }
 
-    @Test(expected = ArrayIndexOutOfBoundsException.class)
-    public void testStackOverFlow() {
-        Stack s = new Stack();
-        for (int i = 0; i < 11; i++) {
-            s.push(i);
+    @Test
+    public void testMaxArraySize() {
+        Stack<Integer> s = new Stack<>(1024); // Stack.MAX_ARRAY_SIZE
```

Comment:
Make MAX_ARRAY_SIZE public, and use it instead of the constant?

Reply:
This is a problem I have with learning Java. I like your suggestion, and even used it for a while, but Sun by way of Oracle makes me think accessing members directly without a getter is worse than killing babies. Could you expand on that?